### PR TITLE
Visitor can buy subscription

### PR DESCRIPTION
**Pull request by tochman**
#### Description
_As a Visitor
In order to get access to the best articles
I would like to be able to buy a subscription_

Changes proposed in this pull request:

* create sign up and payment form for subscription to the site
* add payment with stripe (stripe-rails gem)
* add stripe credentials
* send confirmation email with temporary password if payment is successful
* redirect registered user to set password in order to become a subscriber 
* add registered and subscriber roles to user model
#### What I have learned working on this feature:
* how to personalize devise methods/views
* how to implement stripe
 
#### Screenshots:
image

### CODE REVIEW

File: Gemfile [hunk 1]
```diff
@@ -14,15 +14,21 @@ gem 'webpacker', '~> 3.5'
...
   gem 'rspec-rails'
   gem 'shoulda-matchers'
+	gem 'email_spec'
   gem 'cucumber-rails', require: false
   gem 'database_cleaner'
+	gem 'chromedriver-helper'
+	gem 'selenium-webdriver'
   gem 'coveralls', require: false
+  gem 'launchy'
+	gem 'stripe-ruby-mock', '~> 2.5.6', :require => 'stripe_mock'
```
Comment (tochman) [inline Gemfile]: Hmm... Please use new hash syntax....

File: app/controllers/application_controller.rb [hunk 2]
```diff
@@ -1,7 +1,8 @@
 class ApplicationController < ActionController::Base
-	before_action :set_locale
+  before_action :set_locale
+  # before_action :redirect_if_password_empty
```
Comment (tochman) [inline app/controllers/application_controller.rb]: Clean this up please

File: app/controllers/confirmations_controller.rb [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+			respond_with_navigational(resource.errors, status: :unprocessable_entity){ render :new }
+		end
+	end
+
+	private
+
+	def after_confirmation_path_for(resource_name, resource)
+		sign_in(resource)
+		edit_user_password_path
+	end
```
Comment (tochman) [inline app/controllers/confirmations_controller.rb]: Are you using this in this flow? If not please remove.

File: app/javascript/packs/application.js [hunk 4]
```diff
@@ -6,6 +6,34 @@
 //
 // To reference this file, add <%= javascript_pack_tag 'application' %> to the appropriate
 // layout file, like app/views/layouts/application.html.erb
+import "../css/tailwind.css"
+import { locale } from "core-js";
```
Comment (tochman) [inline app/javascript/packs/application.js]: ??

File: app/mailers/user_mailer.rb [hunk 5]
```diff
@@ -0,0 +1,3 @@
+class UserMailer < ApplicationMailer
+	default from: 'notifications@thehubnews.com'
+end
```
Comment (tochman) [inline app/mailers/user_mailer.rb]: Do you need this mailer?

File: features/Visitor_can_register_for_an_account.feature [hunk 6]
```diff
@@ -1,13 +1,13 @@
...
-      Then I should see "Welcome! You have signed up successfully."
\ No newline at end of file
+  # Scenario:
+  #   Given I visit the "landing" page
+  #   And I click on "Become a subscriber"
+  #   And I fill in "Email" with "person@mail.com"
+  #   And I fill in "Password" with "password"
+  #   And I fill in "Password confirmation" with "password"
+  #   And I click on "Sign up"
+  #   Then I should see "Welcome! You have signed up successfully."
```
Comment (tochman) [inline features/Visitor_can_register_for_an_account.feature]: You can probably remove the whole file at this stage.

File: features/step_definitions/assertion_steps.rb [hunk 7]
```diff
@@ -33,4 +33,23 @@
...
+
+Then("I should be a subscriber") do
+	@user = User.last
+	expect(@user.subscriber?).to eq true
+end
+
+Then("I should see my temporary password in the email body") do
+	@user = User.last
+	password = @user.temp_password
+	expect(current_email.default_part_body.to_s).to include(password)
```
Comment (tochman) [inline features/step_definitions/assertion_steps.rb]: IMO this file is not formatted correctly and should be addressed

File: features/step_definitions/basic_steps.rb [hunk 8]
```diff
@@ -33,5 +37,19 @@
...
+end
+
+Then("show me the page") do
+  save_and_open_page
+end
+
+
+Then("I fill in {string} with my temp password") do |field|
+  fill_in field, with: @user.temp_password
+end
```
Comment (tochman) [inline features/step_definitions/basic_steps.rb]: Same here. I will not comment on this, but this is telling me that your team communication and your ability to collaborate need to improve.

File: features/step_definitions/assertion_steps.rb (hunk 7 above)
Comment (LauraRe) [reply features/step_definitions/assertion_steps.rb]: looks fine on my editor, tab size 2 like the other files. We are going to ask a coach to look into this with us today.

File: features/step_definitions/stripe_related_steps.rb [hunk 9]
```diff
@@ -0,0 +1,33 @@
...
+  when 'Cardnumber'
+    frame_name = '__privateStripeFrame4'
+    field_name = 'cardnumber'
+  when 'Expiry date'
+    frame_name = '__privateStripeFrame5'
+    field_name = 'exp-date'
+  when 'CVC'
+    frame_name = '__privateStripeFrame6'
+    field_name = 'cvc'
+	end
```
[the remark this answers is not in the document]
Comment (LauraRe) [reply features/step_definitions/stripe_related_steps.rb]: :)

File: features/step_definitions/basic_steps.rb (hunk 8 above)
Comment (LauraRe) [reply features/step_definitions/basic_steps.rb]: done

File: features/Visitor_can_register_for_an_account.feature (hunk 6 above)
Comment (LauraRe) [reply features/Visitor_can_register_for_an_account.feature]: deleted

File: app/mailers/user_mailer.rb (hunk 5 above)
Comment (LauraRe) [reply app/mailers/user_mailer.rb]: deleted

File: app/controllers/confirmations_controller.rb (hunk 3 above)
Comment (LauraRe) [reply app/controllers/confirmations_controller.rb]: done

File: app/controllers/application_controller.rb (hunk 2 above)
Comment (LauraRe) [reply app/controllers/application_controller.rb]: done

File: Gemfile (hunk 1 above)
Comment (LauraRe) [reply Gemfile]: done

File: features/step_definitions/assertion_steps.rb (hunk 7 above)
Comment (LauraRe) [reply features/step_definitions/assertion_steps.rb]: done

Comment (oliverochman) [reply features/step_definitions/assertion_steps.rb]: If you view the file outside of "files changed" the formatting looks good

**Pull request merged**